contracts/test/ReentrancyTokenAttacker.sol

```diff
@@ -0,0 +1,103 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.19;
+
+// ServiceRegistryTokenUtility interface
+interface IServiceRegistryTokenUtility {
+    /// @dev Deposit a token security deposit for the service registration after its activation.
+    /// @param serviceId Service Id.
+    /// @return isTokenSecured True if the service Id is token secured, false if ETH secured otherwise.
+    function activateRegistrationTokenDeposit(uint256 serviceId) external returns (bool isTokenSecured);
+
+    /// @dev Deposits bonded tokens from the operator during the agent instance registration.
+    /// @param operator Operator address.
+    /// @param serviceId Service Id.
+    /// @param agentIds Set of agent Ids for corresponding agent instances opertor is registering.
+    /// @return isTokenSecured True if the service Id is token secured, false if ETH secured otherwise.
+    function registerAgentsTokenDeposit(
+        address operator,
+        uint256 serviceId,
+        uint32[] memory agentIds
+    ) external returns (bool isTokenSecured);
+
+    /// @dev Refunds a token security deposit to the service owner after the service termination.
+    /// @param serviceId Service Id.
+    /// @return securityRefund Returned token security deposit, or zero if the service is ETH-secured.
+    function terminateTokenRefund(uint256 serviceId) external returns (uint256 securityRefund);
+
+    /// @dev Refunds bonded tokens to the operator during the unbond phase.
+    /// @param operator Operator address.
+    /// @param serviceId Service Id.
+    /// @return refund Returned bonded token amount, or zero if the service is ETH-secured.
+    function unbondTokenRefund(address operator, uint256 serviceId) external returns (uint256 refund);
+
+    /// @dev Drains slashed funds to the drainer address.
+    /// @param token Token address.
+    /// @return amount Drained amount.
+    function drain(address token) external returns (uint256 amount);
+}
+
+contract ReentrancyTokenAttacker {
```

Comment:
Token reentrancy attacker to test all the possible token transfer failures and reentrancy